### PR TITLE
Implement match types.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

File: tasty-query/shared/src/main/scala/tastyquery/TypeMatching.scala [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  /** Does `tycon` have a field with type `tparam`? */
+  private def typeparamCorrespondsToField(tycon: Type, tparam: TypeParamInfo)(using Context): Boolean =
+    productSelectorTypes(tycon).exists {
+      case tp: TypeRef => tp.isLocalRef(tparam)
+      case _           => false
+    } /*|| tycon.derivesFrom(defn.PairClass)*/
+  end typeparamCorrespondsToField
+
+  private def productSelectorTypes(tycon: Type)(using Context): Iterator[Type] =
+    Iterator
```
Comment (bishabosha) [inline tasty-query/shared/src/main/scala/tastyquery/TypeMatching.scala:301]: questionable, why not look for the case Accessors?

Apparently this is what scala 3 does though

**Review by bishabosha — APPROVED** [review]
initial work looks good, but there should be tests added for match type cases over generic tuple types and enum singleton cases

File: tasty-query/shared/src/test/scala/tastyquery/SubtypingSuite.scala [hunk 2]
```diff
@@ -830,4 +834,228 @@ class SubtypingSuite extends UnrestrictedUnpicklingSuite:
...
+
+    assertEquiv(matchTypeEnums.appliedTo(ctx.findStaticTerm(s"$MyEnumPrefix.Singleton2").staticRef), defn.StringType)
+      .withRef[MatchTypeEnums[MyEnum.Singleton2.type], String]
+
+    assertEquiv(matchTypeEnums.appliedTo(ParametricTycon.appliedTo(defn.StringType)), defn.StringType)
+      .withRef[MatchTypeEnums[MyEnum.Parametric[String]], String]
+  }
+
+  testWithContext("match-types-with-old-style-enums") {
+    import subtyping.TypesFromTASTy.MyEnum
```
Comment (bishabosha) [inline tasty-query/shared/src/test/scala/tastyquery/SubtypingSuite.scala:1038]:
```suggestion
    import subtyping.TypesFromTASTy.OldStyleEnum
```
plus update the other references

Comment (sjrd) [reply tasty-query/shared/src/test/scala/tastyquery/SubtypingSuite.scala]: Oops, indeed. Fixed.

**Review by bishabosha — APPROVED** [review]

**Pull request merged**